eureka_ml_insights/data_utils/transform.py

```diff
@@ -302,3 +303,40 @@ def transform(self, df: pd.DataFrame, encoding="cl100k_base") -> pd.DataFrame:
             token_count_column = f"{column}_token_count"
             df[token_count_column] = token_count
         return df
+
+@dataclass
+class MajorityVoteTransform:
+    """Applies the majority vote transformation to the 'model_output' column per 'ID'."""
+    
+    # No need for additional parameters as the logic is fixed for majority vote.
+    
+    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Transforms the dataframe by calculating the majority vote of 'model_output' per 'ID'.
+        If the 'model_output' is NaN, it will be handled appropriately.
+
+        Args:
+            df (pd.DataFrame): Input dataframe containing 'ID' and 'model_output'.
+        
+        Returns:
+            pd.DataFrame: Transformed dataframe with majority vote for each 'ID'.
+        """
+        
+        # Function to get the majority vote
+        def majority_vote(group):
+            # Drop NaN values and get the most common value
+            group_non_nan = group.dropna()
+            if not group_non_nan.empty:
+                return group_non_nan.mode().iloc[0]  # Return the mode, the most frequent value
+            else:
+                return pd.NA  # Return NaN if all values are NaN
+                
+        df['majority_vote'] = df.groupby('ID')['model_output'].transform(majority_vote)
+
+        df_not_na = df[df['model_output'].notna()]
+        df_not_na = df_not_na[df_not_na['model_output'] == df_not_na['majority_vote']]
+        
+        df_is_na = df[df['majority_vote'].isna()]
```

Comment:
So for each group of samples, you are replacing the non-nan outputs with majority, and keeping the nan ones as nan? Why not replace all of them with majority? What is the thought process behind this and can we explicitly mention  this logic in the doc string?

Reply:
Switched to the aggregate logic.